src/linear_algebra/smodeq.lean

```diff
@@ -77,3 +77,13 @@ begin
 end
 
 end smodeq
+
+/-- The `add_submonoid` consisting of all elements equal to some element in `N` modulo `p`. -/
+def add_submonoid.smodeq_closure (N : add_submonoid M) (U : submodule R M) : add_submonoid M :=
+{ carrier := {x | ∃ r ∈ N, x ≡ r [SMOD U]},
+  add_mem' := λ x y ⟨rx, hx, ex⟩ ⟨ry, hy, ey⟩, ⟨rx + ry, add_mem hx hy, ex.add ey⟩,
+  zero_mem' := ⟨0, zero_mem _, smodeq.refl⟩ }
```

Comment:
This is just `U.to_add_submonoid ⊔ M` right? We could change the following lemma to use that. But if we generalize `U` from submodule to add_subgroup and change `h : x ≡ y [SMOD U]` to `x - y ∈ U`, then it no longer belongs to this file.

Reply:
Maybe we should instead generalize `smodeq` to accept any subobject with `add_subgroup_class`?

Reply:
Sounds a good idea, and I think any result not involving smul could be generalized. However the "S" in SMOD probably stands for "scalar", so the generalized version should have "S" removed; however the [MOD _] is already taken by nat.modeq ...

Reply:
I am inclined to take the liberty to interpret the "S" as `subgroup` and keep it as is.

Reply:
There are four other files that uses smodeq (see below), so it's not gonna be a very small refactor, especially since you are changing the definition (from `submodule.quotient` to [quotient_add_group.subgroup.has_quotient](https://leanprover-community.github.io/mathlib_docs/group_theory/coset.html#quotient_add_group.subgroup.has_quotient)). Notice that the latter allows a noncommutative ambient group, and I wonder if we should also have the multiplicative version ...
```
src/linear_algebra/adic_completion.lean
src/ring_theory/henselian.lean
src/ring_theory/discrete_valuation_ring.lean
src/number_theory/padics/padic_integers.lean 
```

Reply:
It seems there isn't a quotient operation defined to add_subgroup_class. And there doesn't seem exist a def that makes an add_subgroup out of add_subgroup_class either.

Reply:
I'm starting an attempt here: #15915